File: LottoPlayer/app/src/main/java/com/im/daeseong/lottoplayer/Database/DbHandler.java
package com.im.daeseong.lottoplayer.Database;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteException;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;

public class DbHandler {

    private static final String TAG = DbHandler.class.getSimpleName();

    //테이블 이름
    public static final String TABLE_NAME = "Lotto";

    //컬럼 이름
    public static final String COLUMN_RINDEX = "rIndex";
    public static final String COLUMN_DATA = "Date";
    public static final String COLUMN_PART1 = "Part1";
    public static final String COLUMN_PART2 = "Part2";
    public static final String COLUMN_PART3 = "Part3";
    public static final String COLUMN_PART4 = "Part4";
    public static final String COLUMN_PART5 = "Part5";
    public static final String COLUMN_PART6 = "Part6";
    public static final String COLUMN_BONUS = "Bonus";

    private DbHelper dbHelper = null;


    private static DbHandler instance = null;
    public static DbHandler getInstance(Context context) {
        if (instance == null) {
            instance = new DbHandler(context);
        }
        return instance;
    }

    private DbHandler(Context context) {
        dbHelper = DbHelper.getInstance(context);
    }

    public void addLotto(Lotto lotto){

        if(dbHelper == null){
            return;
        }
        SQLiteDatabase db = dbHelper.getWritableDatabase();

        try {

            ContentValues values = new ContentValues();
            values.put(COLUMN_RINDEX, lotto.getrIndex());
            values.put(COLUMN_DATA,   lotto.getDate());
            values.put(COLUMN_PART1,  lotto.getPart1());
            values.put(COLUMN_PART2,  lotto.getPart2());
            values.put(COLUMN_PART3,  lotto.getPart3());
            values.put(COLUMN_PART4,  lotto.getPart4());
            values.put(COLUMN_PART5,  lotto.getPart5());
            values.put(COLUMN_PART6,  lotto.getPart6());
            values.put(COLUMN_BONUS,  lotto.getBonus());
            db.insert(TABLE_NAME, null, values);

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {
            if(db != null){
                db.close();
            }
        }
    }

    public void deleteLotto(int rIndex){

        if(dbHelper == null){
            return;
        }
        SQLiteDatabase db = dbHelper.getWritableDatabase();
        String query = "Delete FROM Lotto where rIndex = " + rIndex;

        try{

            db.execSQL(query);

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {
            if(db != null){
                db.close();
            }
        }
    }

    public ArrayList<Lotto> getLotto() {

        if(dbHelper == null){
            return null;
        }
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        if(db == null){
            return null;
        }

        ArrayList<Lotto> list = new ArrayList<>();

        String query = "SELECT * FROM " + TABLE_NAME + " ORDER BY rIndex DESC";
        Cursor cursor = db.rawQuery(query, null);

        try{

            while (cursor.moveToNext()){
                int rIndex = cursor.getInt(0);
                String Date = cursor.getString(1);
                int Part1 = cursor.getInt(2);
                int Part2 = cursor.getInt(3);
                int Part3 = cursor.getInt(4);
                int Part4 = cursor.getInt(5);
                int Part5 = cursor.getInt(6);
                int Part6 = cursor.getInt(7);
                int Bonus = cursor.getInt(8);
                Lotto lotto = new Lotto(rIndex, Date, Part1, Part2, Part3, Part4, Part5, Part6, Bonus);
                list.add(lotto);
            }

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {

            if(cursor != null){
                cursor.close();
                cursor = null;
            }

            if(db != null){
                db.close();
            }
        }
        return list;
    }

    public int getLottoRowCount(){

        int nTotalcount = 0;

        if(dbHelper == null){
            return nTotalcount;
        }
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        if(db == null){
            return nTotalcount;
        }

        String query = "SELECT count(*) FROM Lotto";
        Cursor cursor = db.rawQuery(query, null);

        try {

            if (cursor.moveToFirst()) {
                cursor.moveToFirst();
                nTotalcount = Integer.parseInt(cursor.getString(0));
            }

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {

            if(cursor != null){
                cursor.close();
                cursor = null;
            }

            if(db != null){
                db.close();
            }
        }
        return nTotalcount;
    }

    public int getPartTop(int nPart) {

        int Part = 0;
        int cnt = 0;
        String query = "";

        if(dbHelper == null){
            return Part;
        }
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        if(db == null){
            return Part;
        }

        if(nPart == 1) {
            query = "SELECT Part1, count(*) as cnt FROM Lotto GROUP BY Part1 order by cnt desc limit 1";
        } else if(nPart == 2) {
            query = "SELECT Part2, count(*) as cnt FROM Lotto GROUP BY Part2 order by cnt desc limit 1";
        } else if(nPart == 3) {
            query = "SELECT Part3, count(*) as cnt FROM Lotto GROUP BY Part3 order by cnt desc limit 1";
        } else if(nPart == 4) {
            query = "SELECT Part4, count(*) as cnt FROM Lotto GROUP BY Part4 order by cnt desc limit 1";
        } else if(nPart == 5) {
            query = "SELECT Part5, count(*) as cnt FROM Lotto GROUP BY Part5 order by cnt desc limit 1";
        } else if(nPart == 6) {
            query = "SELECT Part6, count(*) as cnt FROM Lotto GROUP BY Part6 order by cnt desc limit 1";
        }
        Cursor cursor = db.rawQuery(query, null);

        try {

            if (cursor.moveToFirst()) {
                cursor.moveToFirst();
                Part = Integer.parseInt(cursor.getString(0));
                cnt = Integer.parseInt(cursor.getString(1));
                cursor.close();
            } else {
                Part = 0;
                cnt = 0;
            }

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {

            if(cursor != null){
                cursor.close();
                cursor = null;
            }

            if(db != null){
                db.close();
            }
        }
        return Part;
    }

    public ArrayList<LottoTop> getLottoTop() {

        if(dbHelper == null){
            return null;
        }
        SQLiteDatabase db = dbHelper.getReadableDatabase();
        if(db == null){
            return null;
        }

        ArrayList<LottoTop> sortlist = new ArrayList<>();
        ArrayList<LottoTop> list = new ArrayList<>();

        try{

            setLottoArray(list, "SELECT Part1, count(*) as cnt FROM Lotto GROUP BY Part1 order by cnt desc limit 10");
            setLottoArray(list, "SELECT Part2, count(*) as cnt FROM Lotto GROUP BY Part2 order by cnt desc limit 10");
            setLottoArray(list, "SELECT Part3, count(*) as cnt FROM Lotto GROUP BY Part3 order by cnt desc limit 10");
            setLottoArray(list, "SELECT Part4, count(*) as cnt FROM Lotto GROUP BY Part4 order by cnt desc limit 10");
            setLottoArray(list, "SELECT Part5, count(*) as cnt FROM Lotto GROUP BY Part5 order by cnt desc limit 10");
            setLottoArray(list, "SELECT Part6, count(*) as cnt FROM Lotto GROUP BY Part6 order by cnt desc limit 10");

            DescendingObj descending = new DescendingObj();
            Collections.sort(list, descending);

            for(int i = 0; i < list.size(); i++){
                sortlist.add(list.get(i));
                list.remove(i--);
            }
            sortlist.addAll(sortlist.size(),list);

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {
            if(db != null){
                db.close();
            }
        }

        return sortlist;
    }

    private void setLottoArray(ArrayList<LottoTop> list, String query) {

        if(dbHelper == null){
            return;
        }
        SQLiteDatabase db = dbHelper.getReadableDatabase();

        Cursor cursor = db.rawQuery(query, null);

        try{

            while (cursor.moveToNext()){
                int nNumber = cursor.getInt(0);
                int nCount = cursor.getInt(1);
                LottoTop lotto = new LottoTop(nNumber, nCount);
                list.add(lotto);
            }

        } catch (SQLiteException e){
        } catch (Exception e){
        } finally {

            if(cursor != null){
                cursor.close();
                cursor = null;
            }

            if(db != null){
                db.close();
            }
        }
    }

    //내림 차순 정렬
    class DescendingObj implements Comparator<LottoTop> {
        @Override
        public int compare(LottoTop o1, LottoTop o2) {
            return ((Integer)o2.getCount()).compareTo((Integer)o1.getCount());
        }
    }

    //오름 차순 정렬
    class AscendingObj implements Comparator<LottoTop> {
        @Override
        public int compare(LottoTop o1, LottoTop o2) {
            return ((Integer)o1.getCount()).compareTo((Integer)o2.getCount());
        }
    }

}

File: LottoPlayer(kotlin)/app/src/main/java/com/daeseong/lottoplayer/Database/DbHandler.kt
package com.daeseong.lottoplayer.Database

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteException
import java.util.*
import kotlin.collections.ArrayList
class DbHandler private constructor(context: Context) {

    private val tag = DbHandler::class.java.simpleName

    // Table name
    private val TABLE_NAME = "Lotto"

    // Column names
    private val COLUMN_RINDEX = "rIndex"
    private val COLUMN_DATA = "Date"
    private val COLUMN_PART1 = "Part1"
    private val COLUMN_PART2 = "Part2"
    private val COLUMN_PART3 = "Part3"
    private val COLUMN_PART4 = "Part4"
    private val COLUMN_PART5 = "Part5"
    private val COLUMN_PART6 = "Part6"
    private val COLUMN_BONUS = "Bonus"

    private val dbHelper: DbHelper = DbHelper.getInstance(context)!!

    companion object {
        private var instance: DbHandler? = null

        @Synchronized
        fun getInstance(context: Context): DbHandler {
            if (instance == null) {
                instance = DbHandler(context.applicationContext)
            }
            return instance!!
        }
    }

    fun addLotto(lotto: Lotto) {
        val db = dbHelper.writableDatabase

        try {
            val values = ContentValues()
            values.put(COLUMN_RINDEX, lotto.rIndex)
            values.put(COLUMN_DATA, lotto.date)
            values.put(COLUMN_PART1, lotto.part1)
            values.put(COLUMN_PART2, lotto.part2)
            values.put(COLUMN_PART3, lotto.part3)
            values.put(COLUMN_PART4, lotto.part4)
            values.put(COLUMN_PART5, lotto.part5)
            values.put(COLUMN_PART6, lotto.part6)
            values.put(COLUMN_BONUS, lotto.bonus)
            db.insert(TABLE_NAME, null, values)
        } catch (e: SQLiteException) {
        } finally {
            db.close()
        }
    }

    fun isExistData(rIndex: Int): Boolean {
        var bFindData = false

        val db = dbHelper.readableDatabase ?: return bFindData
        val query = "SELECT rIndex FROM Lotto where rIndex = $rIndex"
        var cursor: Cursor? = null

        try {
            cursor = db.rawQuery(query, null)
            if (cursor.moveToFirst()) {
                bFindData = true
            }
        } catch (e: SQLiteException) {
        } finally {
            cursor?.close()
            db.close()
        }

        return bFindData
    }

    fun getData(rIndex: Int): Array<String?>? {
        val sArray = arrayOfNulls<String>(9)

        val db = dbHelper.readableDatabase ?: return null
        val query = "SELECT * FROM Lotto where rIndex = $rIndex"
        var cursor: Cursor? = null

        try {
            cursor = db.rawQuery(query, null)
            while (cursor.moveToNext()) {
                sArray[0] = cursor.getInt(0).toString()
                sArray[1] = cursor.getString(1)
                sArray[2] = cursor.getInt(2).toString()
                sArray[3] = cursor.getInt(3).toString()
                sArray[4] = cursor.getInt(4).toString()
                sArray[5] = cursor.getInt(5).toString()
                sArray[6] = cursor.getInt(6).toString()
                sArray[7] = cursor.getInt(7).toString()
                sArray[8] = cursor.getInt(8).toString()
            }
        } catch (e: SQLiteException) {
        } finally {
            cursor?.close()
            db.close()
        }

        return sArray
    }

    fun deleteLotto(rIndex: Int) {
        val db = dbHelper.writableDatabase
        val query = "DELETE FROM Lotto WHERE rIndex = $rIndex"

        try {
            db.execSQL(query)
        } catch (e: SQLiteException) {
        } finally {
            db.close()
        }
    }

    fun getLotto(): ArrayList<Lotto>? {
        val db = dbHelper.readableDatabase ?: return null
        val list: ArrayList<Lotto> = ArrayList()
        val query = "SELECT * FROM $TABLE_NAME ORDER BY rIndex DESC"
        var cursor: Cursor? = null

        try {
            cursor = db.rawQuery(query, null)
            while (cursor.moveToNext()) {
                val rIndex = cursor.getInt(0)
                val sDate = cursor.getString(1)
                val nPart1 = cursor.getInt(2)
                val nPart2 = cursor.getInt(3)
                val nPart3 = cursor.getInt(4)
                val nPart4 = cursor.getInt(5)
                val nPart5 = cursor.getInt(6)
                val nPart6 = cursor.getInt(7)
                val nBonus = cursor.getInt(8)
                val lotto = Lotto(rIndex, sDate, nPart1, nPart2, nPart3, nPart4, nPart5, nPart6, nBonus)
                list.add(lotto)
            }
        } catch (e: SQLiteException) {
        } finally {
            cursor?.close()
            db.close()
        }

        return list
    }

    fun getLottoRowCount(): Int {
        var nTotalcount = 0

        val db = dbHelper.readableDatabase ?: return nTotalcount
        val query = "SELECT COUNT(*) FROM $TABLE_NAME"
        var cursor: Cursor? = null

        try {
            cursor = db.rawQuery(query, null)
            if (cursor.moveToFirst()) {
                nTotalcount = cursor.getInt(0)
            }
        } catch (e: SQLiteException) {
        } finally {
            cursor?.close()
            db.close()
        }

        return nTotalcount
    }

    fun getPartTop(nPart: Int): Int {
        var part = 0
        var cnt = 0
        var query = ""

        val db = dbHelper.readableDatabase ?: return part

        when (nPart) {
            1 -> query = "SELECT Part1, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part1 ORDER BY cnt DESC LIMIT 1"
            2 -> query = "SELECT Part2, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part2 ORDER BY cnt DESC LIMIT 1"
            3 -> query = "SELECT Part3, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part3 ORDER BY cnt DESC LIMIT 1"
            4 -> query = "SELECT Part4, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part4 ORDER BY cnt DESC LIMIT 1"
            5 -> query = "SELECT Part5, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part5 ORDER BY cnt DESC LIMIT 1"
            6 -> query = "SELECT Part6, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part6 ORDER BY cnt DESC LIMIT 1"
        }

        var cursor: Cursor? = null

        try {
            cursor = db.rawQuery(query, null)
            if (cursor.moveToFirst()) {
                part = cursor.getInt(0)
                cnt = cursor.getInt(1)
            } else {
                part = 0
                cnt = 0
            }
        } catch (e: SQLiteException) {
            // 예외 처리
        } finally {
            cursor?.close()
            db.close()
        }

        return part
    }

    fun getLottoTop(): ArrayList<LottoTop>? {
        val db = dbHelper.readableDatabase ?: return null
        val sortList: ArrayList<LottoTop> = ArrayList()
        val list: ArrayList<LottoTop> = ArrayList()

        try {
            setLottoArray(
                list,
                "SELECT Part1, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part1 ORDER BY cnt DESC LIMIT 10"
            )
            setLottoArray(
                list,
                "SELECT Part2, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part2 ORDER BY cnt DESC LIMIT 10"
            )
            setLottoArray(
                list,
                "SELECT Part3, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part3 ORDER BY cnt DESC LIMIT 10"
            )
            setLottoArray(
                list,
                "SELECT Part4, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part4 ORDER BY cnt DESC LIMIT 10"
            )
            setLottoArray(
                list,
                "SELECT Part5, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part5 ORDER BY cnt DESC LIMIT 10"
            )
            setLottoArray(
                list,
                "SELECT Part6, COUNT(*) AS cnt FROM $TABLE_NAME GROUP BY Part6 ORDER BY cnt DESC LIMIT 10"
            )

            val descending = DescendingObj()
            Collections.sort(list, descending)

            var i = 0
            while (i < list.size) {
                sortList.add(list[i])
                list.removeAt(i--)
                i++
            }
            sortList.addAll(sortList.size, list)

        } catch (e: SQLiteException) {
        } finally {
            db.close()
        }

        return sortList
    }

    private fun setLottoArray(list: ArrayList<LottoTop>, query: String) {
        val db = dbHelper.readableDatabase
        var cursor: Cursor? = null

        try {
            cursor = db!!.rawQuery(query, null)
            while (cursor.moveToNext()) {
                val nNumber = cursor.getInt(0)
                val nCount = cursor.getInt(1)
                val lotto = LottoTop(nNumber, nCount)
                list.add(lotto)
            }
        } catch (e: SQLiteException) {
        } finally {
            cursor?.close()
            db?.close()
        }
    }

    //내림 차순 정렬
    internal class DescendingObj : Comparator<LottoTop?> {
        override fun compare(o1: LottoTop?, o2: LottoTop?): Int {
            return o2!!.count.compareTo(o1!!.count)
        }
    }

    //오름 차순 정렬
    internal class AscendingObj : Comparator<LottoTop?> {
        override fun compare(o1: LottoTop?, o2: LottoTop?): Int {
            return o1!!.count.compareTo(o2!!.count)
        }
    }
}